test/multi_producer_sequencer_tests.cpp

```diff
@@ -0,0 +1,239 @@
+///////////////////////////////////////////////////////////////////////////////
+// Copyright (c) Lewis Baker
+// Licenced under MIT license. See LICENSE.txt for details.
+///////////////////////////////////////////////////////////////////////////////
+
+#include <cppcoro/multi_producer_sequencer.hpp>
+
+#include <cppcoro/config.hpp>
+#include <cppcoro/sequence_barrier.hpp>
+#include <cppcoro/sequence_traits.hpp>
+#include <cppcoro/on_scope_exit.hpp>
+#include <cppcoro/sync_wait.hpp>
+#include <cppcoro/when_all.hpp>
+#include <cppcoro/task.hpp>
+
+#if CPPCORO_OS_WINNT
+# include <cppcoro/io_service.hpp>
+# include <thread>
+# include <chrono>
+#endif
+
+#include <ostream>
+#include "doctest/doctest.h"
+
+DOCTEST_TEST_SUITE_BEGIN("multi_producer_sequencer");
+
+using namespace cppcoro;
+
+#if CPPCORO_OS_WINNT
+
+namespace
+{
+	task<> one_at_a_time_producer(
+		io_service& ioSvc,
+		multi_producer_sequencer<std::size_t>& sequencer,
+		std::uint64_t buffer[],
+		std::uint64_t iterationCount)
+	{
+		co_await ioSvc.schedule();
+
+		const std::size_t bufferSize = sequencer.buffer_size();
+		const std::size_t mask = bufferSize - 1;
+
+		std::uint64_t i = 0;
+		while (i < iterationCount)
+		{
+			auto seq = co_await sequencer.claim_one();
+			buffer[seq & mask] = ++i;
+			sequencer.publish(seq);
+		}
+
+		auto finalSeq = co_await sequencer.claim_one();
+		buffer[finalSeq & mask] = 0;
+		sequencer.publish(finalSeq);
+	}
+
+	task<> batch_producer(
+		io_service& ioSvc,
+		multi_producer_sequencer<std::size_t>& sequencer,
+		std::uint64_t buffer[],
+		std::uint64_t iterationCount,
+		std::size_t maxBatchSize)
+	{
+		co_await ioSvc.schedule();
+
+		const std::size_t bufferSize = sequencer.buffer_size();
+
+		std::uint64_t i = 0;
+		while (i < iterationCount)
+		{
+			const std::size_t batchSize = static_cast<std::size_t>(
+				std::min<std::uint64_t>(maxBatchSize, iterationCount - i));
+			auto sequences = co_await sequencer.claim_up_to(batchSize);
+			for (auto seq : sequences)
+			{
+				buffer[seq % bufferSize] = ++i;
+			}
+			sequencer.publish(sequences);
+		}
+
+		auto finalSeq = co_await sequencer.claim_one();
+		buffer[finalSeq % bufferSize] = 0;
+		sequencer.publish(finalSeq);
+	}
+
+	task<std::uint64_t> consumer(
+		io_service& ioSvc,
```

Comment:
Change this over to use static_thread_pool so that these tests can be run under linux/clang.